core/src/main/scala/io/qbeast/core/model/CubeWeights.scala

```diff
@@ -113,15 +119,50 @@ class CubeWeightsBuilder protected (
         }
       }
     }
-    weights.map {
-      case (cubeId, weightAndCount) if weightAndCount.count == groupCubeSize =>
-        val scale = desiredCubeSize / groupCubeSize
-        CubeNormalizedWeight(cubeId.bytes, NormalizedWeight(weightAndCount.weight) * scale)
-      case (cubeId, weightAndCount) =>
-        CubeNormalizedWeight(
-          cubeId.bytes,
-          NormalizedWeight(desiredCubeSize, weightAndCount.count))
-    }.toSeq
+
+    weights
+      .map {
+        case (cubeId, weightAndCount) if weightAndCount.count == groupCubeSize =>
+          val nw = NormalizedWeight(weightAndCount.weight)
+          (cubeId, CubeInfo(nw, groupCubeSize))
+
+        case (cubeId, weightAndCount) =>
+          val nw = NormalizedWeight(desiredCubeSize, weightAndCount.count)
+          (cubeId, CubeInfo(nw, weightAndCount.count))
+      }
+  }
+
+  /**
+   * Populate cube tree sizes and parent NormalizedWeights for all existing cubes in the tree
+   * in a bottom-up fashion
+   * @param cubeMap local tree
+   * @return
+   */
+  def populateTreeSizeAndParentWeight(
+      cubeMap: mutable.Map[CubeId, CubeInfo]): mutable.Map[CubeId, CubeInfo] = {
+    // Compute tree sizes and parent weights from bottom up.
+    val levelCubes = cubeMap.keys.groupBy(_.depth)
+    val minLevel = levelCubes.keys.min
+    val maxLevel = levelCubes.keys.max
+    (maxLevel until minLevel by -1) foreach { level =>
+      levelCubes(level)
+        .groupBy(c => c.parent.get)
```

Comment:
Do you really need to group the keys by parent? It seems that grouping by level is enough to ensure that the cubes are processed in the right order. Below every sibling only contributes to it's parent weight, so mixing the cubes from the same level should not be a problem. It would make the algorithm simpler to understand and will reduce a bit the memory consumption. The price is more CubeInfo lookups.